### DISCUSSION TITLE
AtomWithStorage returns undefined on initial render after update to v2

### DISCUSSION THREAD

**Discussion by BennyAlex** [pmndrs/jotai#1803] [General]
Hi,

After updating to v2 the Atoms values are undefined when using the Atom with storage until.
The local storage item is still beeing Set but looks like it's not read correctly.
With v1 it worked as expected.

**Comment by dai-shi** [comment]
Yeah, it's an intentional breaking change and a design choice.
For localStorage case you can create a simpler one on your own.

Related discussions: #1737, #1784, #1791

**Reply by BennyAlex** [reply]
I am not sure why this should be intented? 
We have a website and we save something in the local storage for persistence, then, after revisiting the page, the atom value is empty? Doenst makes any sense. Either its a atom wich is in sync with the storage or we have a dynamic loaded atom like from an API or after user interaction

**Reply by wenq1** [reply]
apparently it’s for making SSR happy. Should be easily fixed in application code though. I had the same problem and changed all relevant useEffects to useUpdateEffect

**Reply by BennyAlex** [reply]
kinda bad installing an additional libary just to get an expected behavior from another libary.
If this is some SSR specific 'fix', then it should only apply when the app is running in SSR mode.

**Reply by dai-shi** [reply]
It's also that the code is more readable and maintainable.
We should provide a non-SSR version as a different util.

Would anyone like to try? (We should separate for sync storage and async storage. #1784 is trying to fix for async storage. so, what's missing is for sync storage.)
The implementation is more or less like this:
https://github.com/pmndrs/jotai/blob/0c195b091a73ca031889e3f39adae87d95c1eaeb/docs/guides/persistence.mdx?plain=1#L28-L47
Hmm, I may feel like adding it in a third-party library.

On second thought, I think the code is more or less duplicated.
So, I guess the possible solution for this is:
```js
export function atomWithSyncStorage(key, initialValue, storage = defaultStorage) {
  const valueInStorage = storage.getItem(key);
  return atomWithStorage(key, valueInStorage === NO_STORAGE_VALUE ? initialValue : valueInStorage, storage)
}
```
Hmmmm, I feel like adding it in a third-party library along with `atomWithAsyncStorage`.

**Reply by BennyAlex** [reply]
Thanks you for rethinking about this.
I definitely think is a valid point to have different packages for SSR and SPA.

**Reply by dai-shi** [reply]
Out of curiosity, do you see any real issue with `undefined`? I know it's suboptimal for SPA, but thought it's just a little performance improvement. (which is not the case  for atomWithAsyncStorage though.)

**Reply by BennyAlex** [reply]
![grafik](https://user-images.githubusercontent.com/17402724/222710245-066c5059-b979-40fb-9076-22690dc1a291.png)
Currently I have much boilerplate

**Reply by dai-shi** [reply]
Yes, and my question is what will break with:
```js
const userAtom = atomWithStorage('user', undefined);
// ...
```

**Reply by BennyAlex** [reply]
so for example, in selected domain is save the curretly selected domain by the user, and depending on this domain, I fetch different data. Now if the user revisits the page, its is read as undefined and so I cant make the fetch or have to use a fallback value, but its much nicer if the user sees directly the data he looked at before.

**Reply by dai-shi** [reply]
I don't disagree that this isn't nice in developer experience, but a workaround like this will mostly be fine for the end user, no?
```jsx
const userAtom = atomWithStorage('user', undefined);

  // in component
  const [user] = userAtomValue(userAtom);
  return user ? <User user={user}> : null;
```

So, I'm thinking `atomWithStorage` works for all cases, and `atomWithSyncStorage` for SPA is for further optimization.

**Reply by BennyAlex** [reply]
I am not sure what userAtomValue is and how passing undefined is fixing the inital value problem?

**Reply by dai-shi** [reply]
No, it doesn't. My question is if the initial value problem is a real problem to end users.

**Reply by BennyAlex** [reply]
We have to write more boilerplate code and also as a new user, I would assume the value is always there when I save it to local storage

**Reply by wootra** [reply]
I am a little confused about atomWithStorage. can't we just add the 3rd argument as localStorage?
```
const userAtom = atomWithStorage('user', localStorage.getItem('user') || initValue, localStorage);
```
